### PR TITLE
Reimplement PlayerEditBookEvent

### PR DESCRIPTION
**Pull request by willies952002**
Also have a config to enable/disable kicking on invalid data.

### CODE REVIEW

File: Spigot-Server-Patches/0144-Reimplement-PlayerEditBookEvent.patch [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+                         PlayerConnection.LOGGER.error("Couldn\'t set command block", exception4);
++                        if (com.destroystokyo.paper.PaperConfig.invalidDataKick) this.disconnect("Invalid Command Block Data!"); // paper
+                     } finally {
+                         packetdataserializer.release();
+                     }
+@@ -2435,6 +2444,7 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
+                                 }
+                             } catch (Exception exception5) {
+                                 PlayerConnection.LOGGER.error("Couldn\'t set beacon", exception5);
++                                if (com.destroystokyo.paper.PaperConfig.invalidDataKick) this.disconnect("Invalid Veacon Data!"); // paper
```
Comment (Black-Hole) [inline Spigot-Server-Patches/0144-Reimplement-PlayerEditBookEvent.patch]: Veacons :exclamation:

Comment (willies952002) [inline Spigot-Server-Patches/0144-Reimplement-PlayerEditBookEvent.patch]: welp... that's a typo :P

File: Spigot-Server-Patches/0144-Reimplement-PlayerEditBookEvent.patch [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+                         PlayerConnection.LOGGER.error("Couldn\'t set command block", exception4);
++                        if (com.destroystokyo.paper.PaperConfig.invalidDataKick) this.disconnect("Invalid Command Block Data!"); // paper
+                     } finally {
+                         packetdataserializer.release();
+                     }
+@@ -2435,6 +2444,7 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
+                                 }
+                             } catch (Exception exception5) {
+                                 PlayerConnection.LOGGER.error("Couldn\'t set beacon", exception5);
++                                if (com.destroystokyo.paper.PaperConfig.invalidDataKick) this.disconnect("Invalid Beacon Data!"); // paper
```
Comment (kashike) [inline Spigot-Server-Patches/0144-Reimplement-PlayerEditBookEvent.patch]: I disagree with this, an exception does not mean invalid data was received.

File: Spigot-Server-Patches/0144-Reimplement-PlayerEditBookEvent.patch [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+                         PlayerConnection.LOGGER.error("Couldn\'t set command block", exception3);
++                        if (com.destroystokyo.paper.PaperConfig.invalidDataKick) this.disconnect("Invalid Command Block Data!"); // paper
+                     } finally {
+                         packetdataserializer.release();
+                     }
+@@ -2410,6 +2418,7 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
+                         }
+                     } catch (Exception exception4) {
+                         PlayerConnection.LOGGER.error("Couldn\'t set command block", exception4);
++                        if (com.destroystokyo.paper.PaperConfig.invalidDataKick) this.disconnect("Invalid Command Block Data!"); // paper
```
Comment (kashike) [inline Spigot-Server-Patches/0144-Reimplement-PlayerEditBookEvent.patch]: I disagree with this, an exception does not mean invalid data was received.

File: Spigot-Server-Patches/0144-Reimplement-PlayerEditBookEvent.patch [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+                     PlayerConnection.LOGGER.error("Couldn\'t sign book", exception1);
++                    if (com.destroystokyo.paper.PaperConfig.invalidDataKick) this.disconnect("Invalid Book Data!"); // paper
+                     return;
+                 } finally {
+                     packetdataserializer.release();
+@@ -2341,6 +2348,7 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
+                         }
+                     } catch (Exception exception3) {
+                         PlayerConnection.LOGGER.error("Couldn\'t set command block", exception3);
++                        if (com.destroystokyo.paper.PaperConfig.invalidDataKick) this.disconnect("Invalid Command Block Data!"); // paper
```
Comment (kashike) [inline Spigot-Server-Patches/0144-Reimplement-PlayerEditBookEvent.patch]: I disagree with this, an exception does not mean invalid data was received.

File: Spigot-Server-Patches/0144-Reimplement-PlayerEditBookEvent.patch [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+                             itemstack1.a("pages", (NBTBase) nbttaglist);
+                             itemstack1.setItem(Items.WRITTEN_BOOK);
++                            CraftEventFactory.handleEditBookEvent(player, itemstack1); // Paper
+                         }
+ 
+                         return;
+                     }
+                 } catch (Exception exception1) {
+                     PlayerConnection.LOGGER.error("Couldn\'t sign book", exception1);
++                    if (com.destroystokyo.paper.PaperConfig.invalidDataKick) this.disconnect("Invalid Book Data!"); // paper
```
Comment (kashike) [inline Spigot-Server-Patches/0144-Reimplement-PlayerEditBookEvent.patch]: I disagree with this, an exception does not mean invalid data was received.

**Comment by willies952002** [comment]
@kashike Those error messages were in CB up until [this commit](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/commits/c194444207ae8543dc4ec4669a6e0d446f41d3b6).

**Comment by kashike** [comment]
Doesn't mean they should be re-added.

**Comment by willies952002** [comment]
@kashike that's what the configuration option is for.

File: Spigot-Server-Patches/0144-Reimplement-PlayerEditBookEvent.patch (hunk 2 above)
Comment (aikar) [inline Spigot-Server-Patches/0144-Reimplement-PlayerEditBookEvent.patch]: If an error that never happens under normal conditions triggers, sounds reasonable to do...

**Pull request merged**